lms/djangoapps/courseware/views/views.py

```diff
@@ -1765,7 +1767,7 @@ def render_xblock(request, usage_key_string, check_if_enrolled=True):
         if ancestor_sequence_block:
             # If the SequenceModule feels that gating is necessary, redirect
             # there so we can have some kind of error message at any rate.
-            if ancestor_sequence_block.descendants_are_gated():
+            if ancestor_sequence_block.descendants_are_gated() and not masquerading_as_specific_student:
```

Comment:
This feels like a check that ought to belong inside `descendants_are_gated`. It's similar to other checks inside there (checks `runtime.user_is_staff` -- which I assume is checking the resulting user, not the original user?), and it feels semantically correct for `descendants_are_gated` to have ownership of all the cases where the user is gated or not, rather than pushing some of that context onto its callers.

You could either have the method query `crum` for the current request/user or maybe just pass it in? And then call `is_masquerading_as_specific_student` inside `descendants_are_gated`. I dunno how gross it is for xblocks to know about users/requests. There's a UserService for getting user data, but it doesn't seem to include masquerade info and seems to intentionally obscure access to the original django user. So maybe it is gross for xblocks to know much about user objects. But this method is a weird one and is only called from this one view - maybe it ought not live on the sequence xblock in the first place? Hmm...

Elsewhere in `seq_module`, it uses passed in context to get masquerade info. Like `context.get('specific_masquerade', False)` -- maybe just pass `masquerading_as_specific_student` down to the gate module and have it early exit, like it does for `is_staff_user`. Which would end up as a tiny small code distinction in practice, but feels like a better API boundary to me to push that check down into the method.

Reply:
That makes sense to me, I've updated the code to pass in context to `descendants_are_gated`. I opted for context in case we need to pass more information in the future, even though right now we are just passing the masquerading info in.